src/scmrepo/git/backend/pygit2.py

```diff
@@ -312,6 +313,18 @@ def resolve_commit(self, rev: str) -> "GitCommit":
             [str(parent) for parent in commit.parent_ids],
         )
 
+    def remove(self, paths: Union[str, Iterable[str]], **kwargs):
+        if isinstance(paths, str):
+            paths = [paths]
+        self.repo.index.remove_all(paths)
+        self.repo.index.write()
+        for path in paths:
+            # TODO: Handle pathspecs
+            if os.path.isfile(path):
+                os.remove(path)
+            elif os.path.isdir(path):
+                rmtree(path, ignore_errors=True)
```

Comment:
It's not safe to use `rmtree` here as it will remove any untracked files in that directory tree. This is also why dulwich remove doesn't support dirs and only takes explicit file paths (since the git index does not actually contain directory paths).

Handling directories requires walking the index + workspace trees and diffing them (so we only remove file that exist in both). When you pass a dir into CLI git, it walks that tree prefix to get the explicit file paths, uses those file paths for the index + worktree remove operations, and then checks to see if the parent dir for any removed files is completely empty afterwards (and then does `rmdir` if empty, otherwise it leaves them alone)

Do we need to support a full `rm` from the workspace or can we limit this to only removing from index (equivalent to `git rm --cached`)?

(I guess the question is how are you intending to use this in DVC?)

Reply:
If you do need workspace remove support, in the pygit2 backend we should have
```py
def remove(..., cached: bool = False):
    if not cached:
        raise NotImplementedError
    ...  # index remove behavior
```

And then handle both the index-only and index+worskpace `cached=True/False` cases in dulwich+gitpython remove.

Reply:
Also, if we intend to support directory removal we should be requiring `recursive=True` the same as CLI git (but I would suggest we initially only support explicit paths similar to dulwich for this)

Reply:
> (I guess the question is how are you intending to use this in DVC?)

I thought I might have needed it but I don't. 
Ended up adding it for the sake of covering the git API but it is probably not worth it to add it just for the sake of it until it's actually required by some upstream project